### PR TITLE
Feature group keep

### PR DESCRIPTION
**Pull request by IgorOsiyuk**
image
add a button that opens a modal window where the user is able to choose keeps

image
create the feature that allows choose keeps for group up

image
Set up a button for saving user's choice. App show this button only for 2 and more chosen keeps

image
create a modal window where user can print the name of his new folder

image
create folders block and current block of folder that contains keeps.

image
user can't drag keeps in folders

### CODE REVIEW

File: google_keeps/src/components/keeps/Keep.jsx [hunk 1]
```diff
@@ -1,9 +1,10 @@
-import React, { useRef } from "react";
+import React, { useEffect, useRef } from "react";
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/Keep.jsx:1]: you imports from 'react' twice

File: google_keeps/src/components/keeps/Keep.jsx [hunk 2]
```diff
@@ -26,21 +27,21 @@ const ControlPanel = styled.div`
...
   animation-name: ${keepShow};
   animation-duration: 1s;
-  animation-fill-mode: forwards;
   position: relative;
   left: 0;
   top: 0;
   transition-duration: 0.218s;
   transition-property: top, left;
+  ${({ isGrouping, isPoint }) => `
+     z-index: ${!isGrouping ? "auto" : "999"};
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/Keep.jsx:39]: create an object in theme to indicate elevation

File: google_keeps/src/components/keeps/KeepsContent.jsx [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+const KeepsContent = () => {
+  const { ValidationInput, ValidationForm } = validationBlocks;
+  const { required } = validationData;
+
+  const [isDisabled, setDisabled] = useState(true);
+  const [isGrouping, setGrouping] = useState(false);
+  const [isFolderName, setFolderName] = useState(false);
+  const [folderTitle, setFolderTitle] = useState({
+    title: "",
+  });
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsContent.jsx:67]: you can just use string for it

File: google_keeps/src/components/keeps/KeepsContent.jsx [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+  const createFolder = useCallback(() => {
+    setGrouping(false);
+    setFolderName(true);
+  }, []);
+
+  const onChange = useCallback((e) => {
+    const { name, value } = e.target;
+    setFolderTitle({
+      [name]: value,
+    });
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsContent.jsx:86]: `setFolderTitle(value)`

File: google_keeps/src/components/keeps/KeepsContent.jsx [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+  }, []);
+
+  const pushFolderList = useCallback(
+    (id) => {
+      if (chosenKeepsList.includes(id)) {
+        let index = chosenKeepsList.indexOf(id);
+        let semiArr = [...chosenKeepsList];
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsContent.jsx:93]: use `const` instead

File: google_keeps/src/components/keeps/KeepsContent.jsx [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+      e.preventDefault();
+      formRef.current.validateAll();
+      if (!btnRef.current.context._errors.length) {
+        dispatch(CreateFolderRequest(folderTitle.title, chosenKeepsList)).then(
+          (res) => {
+            if (res) {
+              setFolderName(false);
+              setFolderTitle({
+                title: "",
+              });
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsContent.jsx:115]: setFolderTitle("")

File: google_keeps/src/components/keeps/KeepsContent.jsx [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+  );
+
+  useEffect(() => {
+    if (chosenKeepsList.length > 1) {
+      setDisabled(false);
+    } else {
+      setDisabled(true);
+    }
+    if (addKeep) {
+      if (chosenKeepsList.length >= 1) {
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsContent.jsx:149]: `if (chosenKeepsList.length)`

File: google_keeps/src/components/keeps/KeepsContent.jsx [hunk 8]
```diff
@@ -0,0 +1,202 @@
...
+
+  const addKeepToFolder = useCallback(
+    (e) => {
+      e.preventDefault();
+      dispatch(AddKeepToFolderRequest(folderId, chosenKeepsList)).then(
+        (res) => {
+          if (res) {
+            setGrouping(false);
+            setChosenKeepsList([]);
+            return;
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsContent.jsx:134]: no need return here

File: google_keeps/src/components/keeps/KeepsFolder/KeepsFolder.jsx [hunk 9]
```diff
@@ -0,0 +1,212 @@
+import React, { useCallback, useState, useRef } from "react";
+import { useDispatch } from "react-redux";
+import styled from "styled-components";
+import Keep, { TitleInput, TextInput, KeepWrap } from "../Keep";
+import { FileAddOutline } from "@styled-icons/evaicons-outline/FileAddOutline";
+import { StyledIconBase } from "@styled-icons/styled-icon";
+import { Save } from "@styled-icons/feather/Save";
+import { ChangeFolderTitleRequest } from "../../../redux/action/user";
+export const FolderBlock = styled.div`
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsFolder/KeepsFolder.jsx:9]: new line

File: google_keeps/src/components/keeps/KeepsFolder/KeepsFolder.jsx [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+`;
+
+const FolderTitle = styled.div`
+  ${({ theme }) => `
+    font-size: ${theme.fontSizes.m};
+    font-weight: ${theme.fontWeight.medium};    
+  `}
+  font-style: italic;
+  ${({ isOpen }) => `
+     z-index: ${isOpen ? "-22" : "1"};
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsFolder/KeepsFolder.jsx:25]: what is that magic number? -22

File: google_keeps/src/components/keeps/KeepsFolder/KeepsFolder.jsx [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+      background-color: ${theme.bgColors.secondary};
+      color:${theme.colors.primary};
+    `}
+  }
+`;
+
+const KeepsFolder = (props) => {
+  const { data, folderTitle, setGrouping, setFolderId, setAddKeep } = props;
+  const [isOpenTitleInput, setOpenTitleInput] = useState(false);
+  const [isChangedTitle, setChangedTitle] = useState(folderTitle.title);
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsFolder/KeepsFolder.jsx:144]: incorrect name  is predict that it boolean value

File: google_keeps/src/components/keeps/KeepsFolder/KeepsFolderList.jsx [hunk 12]
```diff
@@ -0,0 +1,64 @@
+import React from "react";
+import styled from "styled-components";
+import { useSelector } from "react-redux";
+import KeepsFolder, { FolderBlock } from "./KeepsFolder";
+import * as selectors from "../../../selectors";
+
+const FolderWrapper = styled.div`
+  /* padding: 16px; */
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsFolder/KeepsFolderList.jsx:8]: comments?

File: google_keeps/src/components/keeps/KeepsFolder/KeepsFolderList.jsx [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+const FolderBlockContainer = styled.div`
+  display: flex;
+  align-items: flex-start;
+  justify-content: space-between;
+  flex-wrap: wrap;
+
+  ${FolderBlock} {
+    flex: 1 1 calc(25% - 32px);
+    margin: 0 16px;
+    margin-bottom: 16px;
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsFolder/KeepsFolderList.jsx:31]: you can write it as 1 margin

File: google_keeps/src/components/keeps/KeepsFolder/KeepsFolderList.jsx [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+  }
+`;
+
+const KeepsFolderList = (props) => {
+  const { setGrouping, setAddKeep, setFolderId } = props;
+  const groupKeeps = useSelector(selectors.groupKeeps);
+
+  if (!groupKeeps || groupKeeps.length === 0) {
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsFolder/KeepsFolderList.jsx:39]: `if (!groupKeeps || !groupKeeps.length)`

File: google_keeps/src/components/keeps/KeepsList.jsx [hunk 15]
```diff
@@ -39,8 +46,8 @@ const KeepsList = () => {
     return newArray;
   };
 
-  if (keepList.length === 0) {
-    return <h2>у вас нет заметок</h2>;
+  if (!keepList || keepList.length === 0) {
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsList.jsx:49]: `!keepList.length`

File: google_keeps/src/components/keeps/KeepsList.jsx [hunk 16]
```diff
@@ -54,6 +61,11 @@ const KeepsList = () => {
             index={index}
             setFromIndex={setFromIndex}
             setKeepPosition={setKeepPosition}
+            draggable={isGrouping ? false : true}
```
Comment (vladkravc) [inline google_keeps/src/components/keeps/KeepsList.jsx:64]: `draggable={!isGrouping}`

File: google_keeps/src/components/pages/Keeps.jsx [hunk 17]
```diff
@@ -26,8 +30,15 @@ const Keeps = () => {
     <>
       <Container>
         <KeepCreate />
-        <KeepsList />
+        {!isKeepLoading &&
+        (!unGroupKeeps || unGroupKeeps.length === 0) &&
+        (!groupKeeps || groupKeeps.length === 0) ? (
```
Comment (vladkravc) [inline google_keeps/src/components/pages/Keeps.jsx:35]:
```
!unGroupKeeps.length
!groupKeeps.length
```

File: google_keeps/src/components/shared/theme.jsx [hunk 18]
```diff
@@ -12,32 +12,34 @@ const theme = {
...
+    fourth: "#dedede",
     errors: "#a90606",
     warning: "#feefc3",
     overlay: "rgba(0, 0, 0, 0.5)",
   },
-  border:{
-    primary:"#e0e0e0"
+  border: {
+    primary: "#e0e0e0",
+    secondary: "#e0e0e0",
```
Comment (vladkravc) [inline google_keeps/src/components/shared/theme.jsx:22]: why are they similar?

File: google_keeps/src/redux/action/user.jsx [hunk 19]
```diff
@@ -127,3 +135,71 @@ export function UpdateKeepPositionRequest(id, data) {
     );
   };
 }
+
+export function CreateFolder() {
+  return {
+    type: actionType.CREATE_FOLDER,
+  };
+}
+
```
Comment (vladkravc) [inline google_keeps/src/redux/action/user.jsx:144]: You can move action creators in separate file

**Review by vladkravc — CHANGES_REQUESTED** [review]

File: google_keeps/src/components/keeps/Keep.jsx (hunk 2 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/Keep.jsx:39]: - [+]

File: google_keeps/src/components/keeps/KeepsContent.jsx (hunk 3 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/KeepsContent.jsx:67]: - [+]

File: google_keeps/src/components/keeps/KeepsContent.jsx (hunk 6 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/KeepsContent.jsx:115]: - [+]

File: google_keeps/src/components/shared/theme.jsx (hunk 18 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/shared/theme.jsx:22]: - [+]

File: google_keeps/src/components/pages/Keeps.jsx (hunk 17 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/pages/Keeps.jsx]: - [+]

File: google_keeps/src/components/keeps/KeepsList.jsx (hunk 15 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/KeepsList.jsx]: - [+]

File: google_keeps/src/components/keeps/KeepsFolder/KeepsFolderList.jsx (hunk 12 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/KeepsFolder/KeepsFolderList.jsx]: - [+]  
- remove comments but block don't remove.

File: google_keeps/src/components/keeps/KeepsFolder/KeepsFolder.jsx (hunk 11 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/KeepsFolder/KeepsFolder.jsx]: - [+]

File: google_keeps/src/components/keeps/KeepsFolder/KeepsFolder.jsx (hunk 10 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/KeepsFolder/KeepsFolder.jsx]: - [+]

File: google_keeps/src/components/keeps/KeepsFolder/KeepsFolder.jsx (hunk 9 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/KeepsFolder/KeepsFolder.jsx]: - [+]

File: google_keeps/src/components/keeps/KeepsContent.jsx (hunk 8 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/KeepsContent.jsx]: - [+]

File: google_keeps/src/components/keeps/KeepsContent.jsx (hunk 7 above)
Comment (IgorOsiyuk) [reply google_keeps/src/components/keeps/KeepsContent.jsx:137]: that feature needs me to understand what type of action should app does.
If user click on "add to folder" btn then I allow him to choose 1 or more cards,
If user click on 'group keeps' btn then I allow him to choose 2 or more cards , but not least

File: google_keeps/src/redux/action/user.jsx (hunk 19 above)
Comment (IgorOsiyuk) [reply google_keeps/src/redux/action/user.jsx]: - [+]

**Review by melonholy — APPROVED** [review]

**Pull request merged**